packages/ts/react-crud/src/locale.ts

```diff
@@ -0,0 +1,106 @@
+
+function getFormatRegex(format: Intl.DateTimeFormat) {
+  const sampleDate = new Date(1234, 5 - 1, 6);
+  const formattedSample = format.format(sampleDate);
+  const pattern = formattedSample
+    .replace('1234', '(?<year>\\d+)')
+    .replace('05', '(?<month>\\d+)')
+    .replace('5', '(?<month>\\d+)')
+    .replace('06', '(?<day>\\d+)')
+    .replace('6', '(?<day>\\d+)');
+
+  return new RegExp(pattern, 'u');
+}
+
+function tryFormatDate(formatter: Intl.DateTimeFormat, value?: string): string {
+  try {
+    return value ? formatter.format(new Date(value)) : '';
+  } catch (e) {
+    return '';
+  }
+}
+
+export class LocaleFormatter {
+  readonly #date: Intl.DateTimeFormat;
+  readonly #localTime: Intl.DateTimeFormat;
+  readonly #localDateTime: Intl.DateTimeFormat;
+  readonly #number: Intl.NumberFormat;
+  readonly #parsePattern: RegExp;
+
+  constructor(locale?: string) {
+    this.#date = new Intl.DateTimeFormat(locale);
+
+    this.#localTime = new Intl.DateTimeFormat(locale, {
+      hour: 'numeric',
+      minute: 'numeric',
+    });
+
+    this.#localDateTime = new Intl.DateTimeFormat(locale, {
+      day: 'numeric',
+      month: 'numeric',
+      year: 'numeric',
+      hour: 'numeric',
+      minute: 'numeric',
+    });
+
+    this.#number = new Intl.NumberFormat(locale, {
+      maximumFractionDigits: 0,
+    });
+
+    this.#parsePattern = getFormatRegex(this.#date);
+  }
+
+  formatDate(value?: string): string;
+  formatDate(value: DatePickerDate): string;
+  formatDate(value?: DatePickerDate | string): string {
+    if (typeof value === 'object') {
+      const { year, month, day } = value;
+      const date = new Date();
+      date.setFullYear(year);
+      date.setMonth(month);
+      date.setDate(day);
+      return this.#date.format(date);
+    }
+
+    return tryFormatDate(this.#date, value);
+  }
+
+  formatLocalTime(value?: string): string {
+    return tryFormatDate(this.#localTime, `2000-01-01T${value}`);
+  }
+
+  formatLocalDateTime(value?: string): string {
+    return tryFormatDate(this.#localDateTime, value);
+  }
+
+  formatNumber(value?: number): string {
+    return value && Number.isFinite(value) ? this.#number.format(value) : '';
+  }
+
+  parse(dateString: string): DatePickerDate | undefined {
+    const match = this.#parsePattern.exec(dateString);
+    const year = Number(match?.groups?.year);
+    const month = Number(match?.groups?.month) - 1;
+    const day = Number(match?.groups?.day);
+
+    // Verify that the parsed date is valid
+    const dateInstance = new Date();
+    dateInstance.setFullYear(year);
+    dateInstance.setMonth(month);
+    dateInstance.setDate(day);
+
+    if (dateInstance.getFullYear() !== year || dateInstance.getMonth() !== month || dateInstance.getDate() !== day) {
+      return undefined;
+    }
+
+    return { year, month, day };
+  }
+}
+
+export function useLocaleFormatter(): LocaleFormatter {
+  return new LocaleFormatter(useContext(LocaleContext));
```

Comment:
May be a minor thing, but can we wrap this into a `useMemo`?

Reply:
Oops, missed this comment. 

I decided that memoizing formatter does not worth it since it is just a simple instance of the class with couple of objects within and does not do any complex calculations. What do you think?

Reply:
It seems like an easy optimization and avoids:
- Doing the regex format extraction on every render
- Changing the `i18n` property of date pickers on every render, which could have unexpected side-effects. Couldn't observe anything from some quick testing, but the component does quite a lot of things when `i18n` changes, so it might be safer to avoid that.

Reply:
Ok, got it. Gonna add it then

Reply:
Done!